include/ginkgo/core/base/utils_helper.hpp

```diff
@@ -56,29 +56,71 @@ namespace gko {
 class Executor;
 
 
-namespace detail {
+/**
+ * This class is used for function parameters in the place of raw pointers.
+ * It can be converted to from raw pointers, shared pointers and unique pointers
+ * of the specified type or any derived type. This allows functions to be called
+ * without having to use gko::lend or calling .get() for every pointer argument.
+ * It probably has no use outside of function parameters, as it is immutable.
+ *
+ * @tparam T  the pointed-to type
+ */
+template <typename T>
+class pointer_param {
```

Comment:
this should only be used as a parameter, thus the `_param` suffix. otherwise people might accidentally use it for variables in their, see the discussion in #179